rfcs/20230420-kernel-cache/README.md

```diff
@@ -0,0 +1,292 @@
+kernel cache. Also, both options allow users to control the caches capacity
+therefore, the only difference between the options is in the flexibility.
+The second option is in fact the first one with advanced controlling
+capabilities.
+
+Given that the first option will address the problem with dynamic shapes
+related creation overhead the proposal is to start with the it and add the
+API from the option two if we see any need in it.
+
+### Option 2: Extending Runtime Dimensions Support
+
+This option suggests extending the existing runtime dimensions support for
+relevant primitives.
+
+Pros:
+* Using the existing mechanism, the primitive cache will automatically work
+for primitives created with runtime dimensions
+* Unlike the kernel cache, with this approach the primitive and primitive descriptor
+is not re-created therefore there is no related overhead
+
+Cons:
+* The users have to modify their code to enable runtime dimensions
+* Some primitives will require to generate multiple kernels so that the
+primitives can be used with runtime dimensions. It may not be always feasible to
+generate a reasonable number of kernels
+* For some kernels it may not be possible to stay performant and support runtime
+dimensions for all dimensions. This means that if user uses incorrect combination
+of runtime dimensions they may get suboptimal performance
+    * There is an option to document all combinations that will not introduce
+    performance regressions however this would be essentially exposing
+    implementation details to the users. This will require us to make sure that
+    the documented combinations are valid for all hardware and implementations
+    which will have maintenance cost
+
+
+### Conclusion
+
+The runtime dimensions feature eliminates most creation overhead (primitive descriptor,
+primitive, kernel). However, the primitives may have to generate multiple kernels to
+support the feature. And if there are too many (or even indefinite amount of) kernels
+to generate then the runtime dimensions cannot be supported or can be supported partially
+which would mean the supported combinations will have to be documented (i.e., exposing
+implementation details).
+
+The kernel cache feature eliminates only kernel creation overhead that is usually the
+most significant one. However, it will work for all primitives whenever the kernels can be
+re-used. With this feature there is no need to generate all possible kernels, only those
+that are used in the workload will be generated and re-used.
+
+As we can see there is not an ideal solution to the original problem. The conclusion is
+that the runtime dimensions feature work better for optimizations very specific cases
+but provide does the best job at eliminating any creation overhead. While the kernel
+cache feature covers as many cases as possible providing the best result overall.
+
+The proposal is to proceed with the kernel cache feature.
```

Comment:
> The conclusion is that the runtime dimensions feature work better for optimizations very specific cases but provide does the best job at eliminating any creation overhead.

May I know how the runtime dimension feature will be like in the long term? Will other primitives (besides matmul primitive) support this at least at functionality level? 

Personally from Graph API's view, what direction will be converged of dynamic shape support in primitive API matters since it may affect the integration method in frameworks and the design of DNNL backend. With kernel cache, frameworks provide concrete shapes to Graph API and DNNL backend uses it to create primitives. The overhead of compiled partition and primitive creation cannot be reduced. While with extended runtime dimension support, frameworks can pass dynamic dimensions to Graph API and DNNL backend transforms them to runtime dimensions and then create primitives. The compiled partition and primitive creation happens only once.

Reply:
> May I know how the runtime dimension feature will be like in the long term? Will other primitives (besides matmul primitive) support this at least at functionality level?

The semantics of the feature is already defined so I don't think we would like to change it. As for the extending the feature to other primitives, I think this is a big question mark because we don't know if we can enable the feature while maintaining a reasonable performance of the kernels. This is why we have decided to start with the implicit kernel cache (option 2) to give the CPU and GPU teams enough flexibility to experiment with kernel reusability. Depending on the outcome of the experiments we will decide what we should do next.

> The compiled partition and primitive creation happens only once.

With the kernel cache most of the overhead during the primitive creation step will be eliminated so we need to understand that we don't re-create the whole primitive for different shapes.